# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Parent  5827af52810d241d55758b0efe7fdf1a731bcd41
Feature: Override sender choice of HTML-only if at least one plain text recipient exists.

diff --git a/mailnews/compose/src/nsMsgCompose.cpp b/mailnews/compose/src/nsMsgCompose.cpp
--- a/mailnews/compose/src/nsMsgCompose.cpp
+++ b/mailnews/compose/src/nsMsgCompose.cpp
@@ -4628,16 +4628,18 @@ nsMsgCompose::DetermineHTMLAction(int32_
   // Auto-Detect, used by Auto-Detect to determine the appropriate message
   // delivery format.
 
   // allHtml: All recipients prefer HTML.
   bool allHtml = true;
 
   // allPlain: All recipients prefer Plain Text.
   bool allPlain = true;
+  // onePlain: At least one recipient prefers Plain Text.
+  bool onePlain = false;
 
   // Exit the loop early if allHtml and allPlain both decay to false to save us
   // some work.
   for (int i = 0; i < MAX_OF_RECIPIENT_ARRAY && (allHtml || allPlain); ++i) {
     uint32_t nbrRecipients = recipientsList[i].Length();
     for (uint32_t j = 0; j < nbrRecipients && (allHtml || allPlain); ++j) {
       nsMsgRecipient& recipient = recipientsList[i][j];
       uint32_t preferFormat = nsIAbPreferMailFormat::unknown;
@@ -4665,16 +4667,17 @@ nsMsgCompose::DetermineHTMLAction(int32_
       // the summary recipient scopes of the message accordingly.
       switch (preferFormat) {
         case nsIAbPreferMailFormat::html:
           allPlain = false;
           break;
 
         case nsIAbPreferMailFormat::plaintext:
           allHtml = false;
+          onePlain = true;
           break;
 
         default:  // nsIAbPreferMailFormat::unknown
           allHtml = false;
           allPlain = false;
           break;
       }
     }
@@ -4708,16 +4711,20 @@ nsMsgCompose::DetermineHTMLAction(int32_
   NS_ENSURE_SUCCESS(rv, rv);
 
   // If the action is a known send format, return the value to send in that
   // format. Otherwise, ask the user. Note that the preference may default to 0
   // (Ask), which is not a valid value for the following enum.
   if (action == nsIMsgCompSendFormat::PlainText ||
       action == nsIMsgCompSendFormat::HTML ||
       action == nsIMsgCompSendFormat::Both) {
+    if (onePlain && action == nsIMsgCompSendFormat::HTML) {
+      // Make sure plain text recipient gets their plain text.
+      action = nsIMsgCompSendFormat::Both;
+    }
     *result = action;
     return NS_OK;
   }
 
   // At this point, ask the user.
   *result = nsIMsgCompSendFormat::AskUser;
   return NS_OK;
 }
